macrosynergy/download/dataquery.py

```diff
@@ -33,212 +41,234 @@
-    logger.info(f"Requesting URL: {log_url} , track_id: {track_id}")
+class HeartbeatError(Exception):
+    """Raised when a heartbeat fails."""
 
-    try:
-        with requests.request(
-            method=method,
-            url=url,
-            cert=cert,
-            headers=headers,
-            params=params,
-            **kwargs,
-        ) as r:
-            last_url: str = r.url
-            js, success, msg = valid_response(r=r, track_id=track_id)
-    except requests.exceptions.ChunkedEncodingError as e:
-        logger.error(
-            f"ChunkedEncodingError: {e}," f"URL: {log_url}, track_id: {track_id}"
-        )
-        js, success, msg = None, False, None
-        raise InvalidResponseError(
-            e,
-            f"URL : {log_url}",
-        )
 
-    if not success:
-        logger.error(
-            "Request failed for URL: %s with message: %s and response: %s",
-            last_url,
-            msg,
-            js,
-        )
-        if msg["status_code"] == 401:
-            logger.error(
-                "Invalid credentials. Request failed for URL: %s"
-                "with message: %s and response: %s",
-                last_url,
-                msg,
-                js,
-            )
-            raise AuthenticationError(msg)
+def validate_response(response: requests.Response) -> dict:
+    """
+    Validates a response from the API. Raises an exception if the response
+    is invalid (e.g. if the response is not a 200 status code).
 
-    return js, success, last_url, msg
+    :param <requests.Response> response: response object from requests.request().
 
+    :return <dict>: response as a dictionary. If the response is not valid,
+        this function will raise an exception.
 
-class CertAuth(object):
-    """Certificate Authentication.
+    :raises <InvalidResponseError>: if the response is not valid.
+    :raises <AuthenticationError>: if the response is a 401 status code.
+    :raises <KeyboardInterrupt>: if the user interrupts the download.
+    """
+
+    error_str = (
+        f"Response : {response}\n"
+        f"Requested URL: {response.request.url}\n"
+        f"Response status code: {response.status_code}\n"
+        f"Response headers: {response.headers}\n"
+        f"Response text: {response.text}\n"
+        f"Timestamp (UTC) : {datetime.utcnow().isoformat()}; \n"
+    )
+    # TODO : Use response.raise_for_status() as a better way to check for errors
+    if response.status_code == 200:
+        try:
+            response_dict = response.json()
+            if response_dict is None:
+                raise InvalidResponseError(f"Response is empty.\n{error_str}")
+            return response_dict
+        except Exception as exc:
+            if isinstance(exc, KeyboardInterrupt):
+                raise exc
+
+            raise InvalidResponseError(
+                error_str + f"Error parsing response as JSON: {exc}"
```

Comment:
We are raising a general exception, but relating it to the error parsing JSON - shouldn't we filter this out by looking for the specific `JSONDecodeError`?

Reply:
That is right. I tried not to filter the exception as the only relevant exceptions from `response_dict = response.json()` are some form of JSONDecodeError. I will run some mock tests on this with multiple types of errors and move ahead from there.